lib/bdev/nvme/blockdev_nvme.c

```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
 }
 
 static void
-blockdev_nvme_timeout_cb(void *cb_arg, struct spdk_nvme_ctrlr *ctrlr,
-			 struct spdk_nvme_qpair *qpair, uint16_t cid)
+abort_cb(void *arg, const struct spdk_nvme_cpl *cpl)
+{
+	/*
+	 * abort completion callback function. Currently, there is nothing to do
+	 * struct spdk_nvme_ctrlr *ctrlr = arg;
+	 */
+}
+
+static void
+blockdev_nvme_timeout_cb(struct spdk_nvme_ctrlr *ctrlr,
```

Comment:
I just did a bunch of tests on an Intel P3700 and learned a little bit. For the ADMIN commands I tried (log pages, set features, etc.) they appear to execute serially. The device does not pick up the next ADMIN command from the queue until the previous one has completed. That means aborting an ADMIN command is impossible (because abort is an ADMIN command and won't be seen until the previous commands complete). This is of course implementation specific, and I didn't send every possible ADMIN command, but it's evidence that aborting ADMIN commands is not going to be a reliable strategy for error recovery across all NVMe devices.

Aborts on I/O commands, on the other hand, work fine. Therefore, I think the right strategy is to send an ABORT when an I/O command times out. If any ADMIN command times out, reset the controller.